### PR TITLE
Add S3 and Wasabi as storage providers

### PR DESCRIPTION
**Pull request by takielias**
### **Description:**
This PR introduces Wasabi as a new storage provider within the Vito Deploy system. Wasabi is a cloud storage service that offers S3-compatible object storage. This addition aligns with the existing storage providers such as FTP and Dropbox, providing users with more flexible and cost-effective storage options.

### **Changes Made:**

- **Modified Files:**
  - **`app/Enums/StorageProvider.php`**: Added Wasabi as a new enum value to support selection within the system.
  - **`config/core.php`**: Updated configuration to include Wasabi, enabling it to be properly recognized and managed by the system.

- **New Files:**
  - **`app/StorageProviders/Wasabi.php`**: Implements the Wasabi-specific logic for managing storage operations, utilizing Wasabi’s S3-compatible API.
  - **`app/SSH/Storage/Wasabi.php`**: Manages the interaction between the SSH layer and Wasabi, ensuring that commands are executed correctly in a remote environment.
  - **`app/SSH/Storage/scripts/wasabi/download.sh`**: Bash script for downloading files from Wasabi storage.
  - **`app/SSH/Storage/scripts/wasabi/upload.sh`**: Bash script for uploading files to Wasabi storage.

- **New Tests:**
  - **`tests/Feature/StorageProvidersTest.php`**: Added test cases for Wasabi to ensure it can be correctly configured, connected, and used for file operations, similar to existing storage providers.

image

### **Reason for Changes:**
- To provide an additional storage option that leverages Wasabi’s S3 compatibility, offering users more flexibility in choosing a cloud storage provider.
- To extend the functionality of the Vito Deploy system to accommodate different storage needs, particularly for users looking for a cost-effective solution.

### **Testing:**
- **Unit Tests**: Updated and added test cases in `StorageProvidersTest.php` to ensure that Wasabi can be correctly configured, connected, and used for file operations.
- **Integration Tests**: Conducted integration testing to verify that Wasabi interacts smoothly with the existing deployment workflows.
- **Manual Testing**: Performed manual checks to ensure that files can be uploaded, downloaded, and deleted from Wasabi storage without issues.

### **Impacted Areas:**
- Storage provider management, specifically within the enum and configuration files.
- Script execution for file operations via SSH, tailored for Wasabi’s S3-compatible API.

### **Notes:**
- This change is backward compatible and does not interfere with existing storage providers.
- Users will need to input their Wasabi credentials and specify the correct region when configuring the storage provider.

### **Reviewer Notes:**
- Please review the Wasabi-specific logic in both the `StorageProviders` and `SSH` layers.
- Verify that the new scripts for Wasabi (`download.sh` and `upload.sh`) are correctly integrated and functioning as expected.
- Review the new test cases in `StorageProvidersTest.php` to ensure they cover all necessary scenarios.
- Feedback on code structure, especially in handling the S3-compatible API, would be valuable.

### CODE REVIEW

**Comment by saeedvaziry** [comment]
Thanks for the PR @takielias.

Why not making just a S3 driver instead of wasabi wich uses S3 under the hood?

**Comment by gaetan-hexadog** [comment]
Hi @takielias ,

thanks a lot for this PR. I don't know wasabi but I see its a S3-compatible object storage. Why not calling it "S3 Storage" instead of Wasabi ? It should work with any S3-compatible object storage provider right ?

On my side I'm not using Wasabi but Tebi.io and I'm looking forward to be able to connect my vito instance to it. So naming it with a more generic name could help do not confuse user who want to connect any S3-compatible object storage.

Best regards

**Comment by takielias** [comment]
> Thanks for the PR @takielias.
> 
> Why not making just a S3 driver instead of wasabi wich uses S3 under the hood?

@saeedvaziry Thank you for the feedback! I decided to start with Wasabi because I've been using it extensively for the past 4 years and am more familiar with its nuances. There were also some conditional differences in implementing the AWS CLI for Wasabi that made it a more natural starting point for me. I agree that a more generalized S3 driver would be a great addition, and I'll definitely consider working on that in the near future.

**Comment by takielias** [comment]
Hi @austinkregel,

Thank you for the thoughtful feedback! You're absolutely right—Wasabi is S3-compatible, and the implementation could indeed work with any S3-compatible object storage provider. I initially focused on Wasabi because it’s what I use regularly.

I'll consider refactoring the implementation to make it more universally applicable to any S3-compatible storage in near future. 

Thanks again for the suggestion!

File: app/SSH/Storage/scripts/wasabi/upload.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+# Check if AWS CLI is installed
+if ! command -v aws &> /dev/null
+then
+    echo "AWS CLI is not installed. Installing..."
+
+    # Install AWS CLI v2 with the --update flag
+    sudo curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
```
Comment (saeedvaziry) [inline app/SSH/Storage/scripts/wasabi/upload.sh:9]: will this work on the arm based CPUs?

**Comment by saeedvaziry** [comment]
@takielias I would still go with S3 to build the core and then will extend Wasabi from S3. I can help with the PR to cover both.

File: app/SSH/Storage/scripts/wasabi/upload.sh (hunk 1 above)
Comment (takielias) [reply app/SSH/Storage/scripts/wasabi/upload.sh:9]: It works on x86_64 (64-bit Intel/AMD) CPUs. Need some modification to download ARM-compatible version of the AWS CLI. I'm going to update it.

**Comment by takielias** [comment]
> @takielias I would still go with S3 to build the core and then will extend Wasabi from S3. I can help with the PR to cover both.

Thant would be great. Please.

File: app/SSH/Storage/scripts/wasabi/upload.sh (hunk 1 above)
Comment (takielias) [reply app/SSH/Storage/scripts/wasabi/upload.sh]: https://github.com/vitodeploy/vito/pull/281/commits/c4ff1deebe0ead6e77d9103ea3ceb8f9233fd404 Fixed.

**Comment by takielias** [comment]
@saeedvaziry I'm working on S3. Will share the update soon.

**Comment by takielias** [comment]
Hi @saeedvaziry,

I've updated the PR to incorporate a core S3 compatibility layer with the following changes:

`S3AbstractStorageProvider.php` and `S3AbstractStorage.php`: These files serve as the core for any S3-compatible service.

`HasS3Storage.php`: A trait added to ensure consistent file path management across S3-compatible services.

`Wasabi Storage`: I've refactored the Wasabi storage implementation to extend from this new S3 core, allowing for easier integration of other S3-compatible storage providers in the future.

_I have also update the tests_ `tests/Feature/StorageProvidersTest.php`

image

Please take a look at the code and let me know if you have any suggestions or improvements. I'm open to feedback and eager to make this as robust and flexible as possible.

**Comment by saeedvaziry** [comment]
@takielias the actions are failing

**Comment by takielias** [comment]
> @takielias the actions are failing

@saeedvaziry 

I have fixed the tests.

File: tests/Feature/StorageProvidersTest.php [hunk 2]
```diff
@@ -101,6 +124,154 @@ public function test_cannot_delete_provider(): void
...
+                'region' => 'us-east-1',
+                'bucket' => 'fake-bucket',
+                'path' => '/',
+            ]
+        ]);
+
+        // Mock the S3Client's method `listBuckets`
+        $s3ClientMock = $this->getMockBuilder(S3Client::class)
+            ->disableOriginalConstructor()
+            ->addMethods(['listBuckets'])
```
Comment (saeedvaziry) [inline tests/Feature/StorageProvidersTest.php:217]: addMethods is deprecated

File: tests/Feature/StorageProvidersTest.php [hunk 3]
```diff
@@ -101,6 +124,154 @@ public function test_cannot_delete_provider(): void
...
```
Comment (saeedvaziry) [inline tests/Feature/StorageProvidersTest.php:254]: same here, deprecated

**Comment by saeedvaziry** [comment]
fixes #263

File: tests/Feature/StorageProvidersTest.php (hunk 2 above)
Comment (takielias) [reply tests/Feature/StorageProvidersTest.php]: It's fixed now.

File: tests/Feature/StorageProvidersTest.php (hunk 3 above)
Comment (takielias) [reply tests/Feature/StorageProvidersTest.php]: It's fixed.

**Comment by manusiakemos** [comment]
does it support minio?

**Comment by atak011** [comment]
@manusiakemos 

You don't need to do anything extra for the minio, the s3 engine and the minio engine ( sdks )are compatible with each other.

**Comment by manusiakemos** [comment]
> @manusiakemos
> 
> You don't need to do anything extra for the minio, the s3 engine and the minio engine ( sdks )are compatible with each other.

can you help me because there is no endpoint input here. so i cannot connect to minio
image

**Pull request merged**